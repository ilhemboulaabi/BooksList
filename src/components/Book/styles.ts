import { Box, Button, Card, Typography } from "@mui/material";
import { palette, spacing } from "../../styles/accessors";
import styled from "../../styles/styled-components";
import { plateform } from "../../styles/theme";

export const SCCard = styled(Card)<{isbookdetails?: string}>`
  && {
    display: flex;
    width: fit-content;
    box-shadow: none;
    height: ${(props) => props?.isbookdetails ? '300px' : '200px'};
    width: 400px;
    @media ${plateform.medium} {
      width: 600px;
    }
    @media ${plateform.large} {
      width: 800px;
    }
  }
`;

export const SCButton = styled(Button)`
  && {
    margin-top: auto;
    align-self: flex-start;
    margin-left: ${spacing(2)};
  }
`;

export const SCFavoriteButton = styled(Button)`
  && {
    margin-left: auto;
    margin-top: auto;
    height: 50px;
    color: ${palette("main")("white")};
    background: ${palette("secondary")("main")};
  }
`;

export const SCTypography = styled(Typography)`
  && {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 100px;
    @media ${plateform.medium} {
      max-width: 300px;
    }
    @media ${plateform.large} {
      max-width: 400px;
    }
  }
`;

export const SCBox = styled(Box)`
  && {
    display: flex;
    flex-direction: column;
    width: 100px;
    @media ${plateform.medium} {
      width: 300px;
    }
    @media ${plateform.large} {
      width: 400px;
    }
  }
`;