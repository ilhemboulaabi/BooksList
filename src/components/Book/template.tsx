import React from 'react';
import {
  CardContent,
  Typography,
  LinearProgress,
  Grid,
  CardMedia
} from "@mui/material";
import FavoriteIcon from '@mui/icons-material/Favorite';
import { IBook, IBookDetails } from '../../types';
import { SCBox, SCButton, SCCard, SCFavoriteButton, SCTypography } from './styles';
import { Link } from 'react-router-dom';

const Book = ({ book, loading, isBookDetails }: { book: IBook | IBookDetails, loading: boolean, isBookDetails?: string }) => {

  return (
    loading || !book
      ? <LinearProgress />
      : (
        <Grid item xs={12}>
          <SCCard raised isbookdetails={isBookDetails}>
            <CardMedia
              component="img"
              sx={{ width: isBookDetails ? 225 : 125, maxHeight: isBookDetails ? 300 : 200 }}
              image={isBookDetails ? `https://covers.openlibrary.org/b/id/${(book as IBookDetails).covers && (book as IBookDetails).covers[0]}-L.jpg` : `https://covers.openlibrary.org/b/id/${(book as IBook).cover_i}-L.jpg`}
              alt="Live from space album cover"
            />
            <SCBox>
              <CardContent>
                <SCTypography variant="h5">
                  {book.title}
                </SCTypography>
                {book.subtitle &&
                  <SCTypography variant="h6">
                    {book.subtitle}
                  </SCTypography>
                }
                <Typography variant="subtitle1" color="textSecondary">
                  <b>By:</b> {(book as IBook).author_name}
                </Typography>
                {isBookDetails &&
                  <>
                    <Typography variant="subtitle1" color="textSecondary">
                      <b>Fist Publish Date:</b>{(book as IBookDetails).first_publish_date}
                    </Typography>
                    <Typography variant="subtitle1" color="textSecondary">
                     <b>Description:</b>{(book as IBookDetails).description && (book as IBookDetails).description.value}
                    </Typography>
                  </>
                }
              </CardContent>
              {!isBookDetails &&
                <SCButton size="small" color="primary">
                  <Link to={`${book.key}`}> More...</Link>
                </SCButton>
              }
            </SCBox>
            <SCFavoriteButton variant="contained" startIcon={<FavoriteIcon />}>
              Favorite
            </SCFavoriteButton>
          </SCCard>
        </Grid>
      )
  );
};

export default Book;
