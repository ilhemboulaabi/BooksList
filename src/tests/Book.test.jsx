import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import Book from '../components/Book/template';
import messageData from './mockData/data';

Enzyme.configure({ adapter: new Adapter() });

const setup = () => {
  const props = {
    book: messageData,
    loading: false,
  };

  const wrapper = shallow(<Book {...props} />);

  return {
    props,
    wrapper,
  };
};

describe('Book', () => {
  it('should render self', () => {
    const { wrapper } = setup();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
