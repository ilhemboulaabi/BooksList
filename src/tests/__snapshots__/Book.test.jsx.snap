// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book should render self 1`] = `
<ForwardRef(Grid)
  item={true}
  xs={12}
>
  <Styled(Component)
    raised={true}
  >
    <ForwardRef(CardMedia)
      alt="Live from space album cover"
      component="img"
      image="https://covers.openlibrary.org/b/id/882715-L.jpg"
      sx={
        Object {
          "maxHeight": 200,
          "width": 125,
        }
      }
    />
    <Styled(Component)>
      <ForwardRef(CardContent)>
        <Styled(Component)
          variant="h5"
        >
          The Power of Darkness
        </Styled(Component)>
        <ForwardRef(Typography)
          color="textSecondary"
          variant="subtitle1"
        >
          By 
          Edith Nesbit
        </ForwardRef(Typography)>
      </ForwardRef(CardContent)>
      <Styled(Component)
        color="primary"
        size="small"
      >
        <Link
          to="/works/OL407498W"
        >
           More...
        </Link>
      </Styled(Component)>
    </Styled(Component)>
    <Styled(Component)
      startIcon={<UNDEFINED />}
      variant="contained"
    >
      Favorite
    </Styled(Component)>
  </Styled(Component)>
</ForwardRef(Grid)>
`;
